src/main/java/com/robertx22/mine_and_slash/capability/entity/ResourceTracker.java

```diff
@@ -119,6 +119,32 @@ public float getKeyProgress(String key, ResourceType rt) {
         return byKey == null ? 0f : byKey.getOrDefault(key, 0f);
     }
 
+    /**
+     * Set the exact cursor value for a specific key/resource.
+     * If value <= EPS the key entry is removed.
+     */
+    public void setKeyProgress(String key, ResourceType rt, float value) {
+        if (key == null || key.isEmpty()) return;
+        var byKey = keyProgress.computeIfAbsent(rt, __ -> new java.util.HashMap<>());
```

Comment:
Better to extract this line into a separate method because it repeats several times during this file and changes all other same occurences
It should be something like
```suggestion
        var byKey = getKeyProgressOrCreate(rt);
```